src/report/report.c

```diff
@@ -0,0 +1,212 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <getopt.h>
+
+#include "alloc-util.h"
+#include "build.h"
+#include "dirent-util.h"
+#include "fd-util.h"
+#include "log.h"
+#include "main-func.h"
+#include "sd-varlink.h"
+#include "string-util.h"
+#include "sd-event.h"
+
+#define RUN_SYSTEMD_METRICS_PATH "/run/systemd/metrics/"
+#define MAX_METRICS_SOCKETS 20
+
+typedef struct Context {
+        int n_ops;
+} Context;
+
+static int metrics_on_query_reply(
+                sd_varlink *link,
+                sd_json_variant *parameters,
+                const char *error_id,
+                sd_varlink_reply_flags_t flags,
+                void *userdata) {
+
+        assert(link);
+
+        Context* context = ASSERT_PTR(userdata);
+
+        if (error_id) {
+                bool disconnect;
+
+                disconnect = streq(error_id, SD_VARLINK_ERROR_DISCONNECTED);
+                if (disconnect)
+                        log_info("Disconnected.");
+                else
+                        log_error("Varlink error: %s", error_id);
+
+                (void) sd_event_exit(ASSERT_PTR(sd_varlink_get_event(link)), disconnect ? EXIT_SUCCESS : EXIT_FAILURE);
+                return 0;
+        }
+
+        sd_json_variant_dump(parameters, SD_JSON_FORMAT_PRETTY_AUTO|SD_JSON_FORMAT_COLOR_AUTO, stdout, NULL);
+
+        fflush(stdout);
+
+        if ((flags == 0) && (--context->n_ops <= 0))
```

Comment:
I'm not sure if `flags == 0` is a correct check. You want to ensure that you decrement counter on the last message for each connection. Other flags may still be set.

Reply:
> For all other messages, the it's SD_VARLINK_REPLY_CONTINUES which is 2.

I think you should be checking for this flag specifically rather then doing a coarse `flags == 0`.